### PR TITLE
[Fix #1977] Allow using danger methods with Time/Date subclasses

### LINKED ISSUE #1977: Time.zone suggestion not smart enough
The following code triggers the "use Time.zone.strftime instead" warning:

```ruby
Faker::Time.forward(0).strftime('%H:%M')
```

`Faker` is from the `faker` gem and `Faker::Time.zone` does not exist (why would it) and `Faker::Time.zone.forward(0).zone` returns the timezone abbreviation which does not have a `strftime` method.

```sh
$ rubocop -V
0.32.0 (using Parser 2.2.2.5, running on ruby 2.2.1 x86_64-linux)
```

### PR DESCRIPTION
**Pull request by palkan**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### Bugs fixed
 
 * [#1988](https://github.com/bbatsov/rubocop/issues/1988): Fix bug in `Style/ParallelAssignment` when assigning from `Module::CONSTANT`. ([@rrosenblum][])
+* [#1977](https://github.com/bbatsov/rubocop/issues/1977): Fix bugs in `Rails/Date` and `Rails/TimeZone` when using Time/Date subclasses.
```
Comment (bbatsov) [inline CHANGELOG.md]: I think subclasses is a bit confusing here - the problem is just the same names in other namespaces.

**Comment by bbatsov** [comment]
You'll have to rebase this.

File: lib/rubocop/cop/rails/time_zone.rb [hunk 2]
```diff
@@ -44,9 +44,9 @@ class TimeZone < Cop
                             :to_i, :to_f]
 
         def on_const(node)
-          _module, klass = *node
+          mod, klass = *node
 
-          return unless method_send?(node)
+          return unless (mod.nil? || mod.cbase_type?) && method_send?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/time_zone.rb]: You should probably add a comment about this particular check.

File: lib/rubocop/cop/rails/time_zone.rb [hunk 3]
```diff
@@ -44,9 +44,9 @@ class TimeZone < Cop
                             :to_i, :to_f]
 
         def on_const(node)
-          _module, klass = *node
+          mod, klass = *node
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/time_zone.rb]: mod -> module

File: spec/rubocop/cop/rails/date_spec.rb [hunk 4]
```diff
@@ -10,9 +10,14 @@
 
     %w(today current yesterday tomorrow).each do |day|
       it "registers an offense for Date.#{day}" do
-        inspect_source(cop, "Date.#{day}")
+        inspect_source(cop, "::Date.#{day}")
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/date_spec.rb]: You can have 2 different examples here - with and without `::`.

File: spec/rubocop/cop/rails/time_zone_spec.rb [hunk 5]
```diff
@@ -136,6 +146,13 @@
       inspect_source(cop, "Time.zone.parse('12:00').localtime('+03:00')")
       expect(cop.offenses).to be_empty
     end
+
+    described_class::DANGER_METHODS.each do |a_method|
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/time_zone_spec.rb]: in another commit you can rename `danger` to `dangerous`.

File: lib/rubocop/cop/rails/time_zone.rb (hunk 3 above)
Comment (palkan) [inline lib/rubocop/cop/rails/time_zone.rb]: `module` is reserved keyword

Comment (bbatsov) [inline lib/rubocop/cop/rails/time_zone.rb]: Of course. Silly me.

On Sunday, June 28, 2015, Vladimir Dementyev <notifications@github.com>
wrote:

> In lib/rubocop/cop/rails/time_zone.rb
> <https://github.com/bbatsov/rubocop/pull/1997#discussion_r33421613>:
>
> > @@ -44,9 +44,9 @@ class TimeZone < Cop
> >                              :to_i, :to_f]
> >
> >          def on_const(node)
> > -          _module, klass = *node
> > +          mod, klass = *node
>
> module is reserved keyword
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/1997/files#r33421613>.
>


-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

**Pull request merged**